temporal-sdk/src/main/java/io/temporal/internal/statemachines/UpdateProtocolStateMachine.java

```diff
@@ -67,19 +78,74 @@ enum State {
   private String requestMsgId;
   private long requestSeqID;
   private Request initialRequest;
+  private String messageId;
 
   public static final StateMachineDefinition<State, ExplicitEvent, UpdateProtocolStateMachine>
       STATE_MACHINE_DEFINITION =
           StateMachineDefinition.<State, ExplicitEvent, UpdateProtocolStateMachine>newInstance(
-                  "Update", State.NEW, State.COMPLETED)
+                  "Update", State.NEW, State.COMPLETED_COMMAND_RECORDED)
               .add(
                   State.NEW,
                   ProtocolType.UPDATE_V1,
                   State.REQUEST_INITIATED,
                   UpdateProtocolStateMachine::triggerUpdate)
-              .add(State.REQUEST_INITIATED, ExplicitEvent.ACCEPT, State.ACCEPTED)
-              .add(State.REQUEST_INITIATED, ExplicitEvent.REJECT, State.COMPLETED)
-              .add(State.ACCEPTED, ExplicitEvent.COMPLETE, State.COMPLETED);
+              .add(
+                  State.REQUEST_INITIATED,
+                  ExplicitEvent.ACCEPT,
+                  State.ACCEPTED,
+                  UpdateProtocolStateMachine::sendCommandMessage)
+              .add(
+                  State.ACCEPTED,
+                  CommandType.COMMAND_TYPE_PROTOCOL_MESSAGE,
+                  State.ACCEPTED_COMMAND_CREATED)
+              .add(
+                  State.ACCEPTED_COMMAND_CREATED,
+                  EventType.EVENT_TYPE_WORKFLOW_EXECUTION_UPDATE_ACCEPTED,
+                  State.ACCEPTED_COMMAND_RECORDED)
+              .add(
+                  State.ACCEPTED_COMMAND_RECORDED,
+                  ExplicitEvent.COMPLETE,
+                  State.COMPLETED,
+                  UpdateProtocolStateMachine::sendCommandMessage)
+              .add(
+                  State.COMPLETED,
+                  CommandType.COMMAND_TYPE_PROTOCOL_MESSAGE,
+                  State.COMPLETED_COMMAND_CREATED)
+              .add(
+                  State.COMPLETED_COMMAND_CREATED,
+                  EventType.EVENT_TYPE_WORKFLOW_EXECUTION_UPDATE_COMPLETED,
+                  State.COMPLETED_COMMAND_RECORDED)
+              // Handle the validation failure case
+              .add(State.REQUEST_INITIATED, ExplicitEvent.REJECT, State.COMPLETED_COMMAND_RECORDED)
+              // Handle an edge case when the update handle completes immediately. The state machine
+              // should then expect
+              // to see two protocol command messages back to back then two update events.
+              .add(
+                  State.ACCEPTED,
+                  ExplicitEvent.COMPLETE,
+                  State.COMPLETED_IMMEDIATELY,
+                  UpdateProtocolStateMachine::sendCommandMessage)
```

Comment:
IMO these ones that are doing send command message would be more clear as having the method that does the explicit event do only that, and then having different implementations of `sendCommandMessage` for the different outcomes.

This way, you don't set the `messageId` class field and then rely on it at a distance a moment later, it can avoid being a class field entirely.

Not blocking by any means and I get the Java state machines do this pattern not infrequently but this approach might be a little more clear / safer.